File: cpp/parser.cpp
#include "parser.hpp"
#include <algorithm>

namespace mpp {

Parser::Parser(std::vector<Token> tokens) : tokens_(std::move(tokens)) {}

const Token& Parser::current() const {
    return pos_ >= tokens_.size() ? tokens_.back() : tokens_[pos_];
}

const Token& Parser::advance() {
    const Token& t = current();
    if (pos_ < tokens_.size()) pos_++;
    return t;
}

bool Parser::check(TokenType t) const { return current().type == t; }
bool Parser::check(TokenType a, TokenType b) const {
    TokenType c = current().type;
    return c == a || c == b;
}

const Token& Parser::expect(TokenType t) {
    if (current().type != t)
        throw ParseError("Unexpected token", current());
    return advance();
}

bool Parser::isAssignable(Expr* e) const {
    return dynamic_cast<IdentExpr*>(e) || dynamic_cast<IndexExpr*>(e) || dynamic_cast<MemberExpr*>(e);
}

std::unique_ptr<ProgramNode> Parser::parse() {
    auto prog = std::make_unique<ProgramNode>();
    while (!check(TokenType::Eof)) {
        auto stmt = parseStatement();
        if (stmt) prog->statements.push_back(std::move(stmt));
        if (check(TokenType::Semicolon)) advance();
    }
    return prog;
}

std::unique_ptr<Stmt> Parser::parseStatement() {
    if (check(TokenType::Let) || check(TokenType::Const))
        return parseVarDecl();
    if (check(TokenType::Fn)) return parseFuncDecl();
    if (check(TokenType::Class)) return parseClassDecl();
    if (check(TokenType::If)) return parseIf();
    if (check(TokenType::For)) return parseFor();
    if (check(TokenType::While)) return parseWhile();
    if (check(TokenType::Return)) {
        advance();
        std::unique_ptr<Expr> val;
        if (!check(TokenType::Semicolon) && !check(TokenType::RBrace) && !check(TokenType::Eof))
            val = parseExpr();
        return std::make_unique<ReturnStmt>(std::move(val));
    }
    if (check(TokenType::LBrace)) return parseBlock();

    auto expr = parseExpr();
    if (expr && check(TokenType::Eq) && isAssignable(expr.get())) {
        advance();
        auto target = std::move(expr);
        auto value = parseExpr();
        auto a = std::make_unique<AssignStmt>();
        a->target = std::move(target);
        a->value = std::move(value);
        return a;
    }
    if (expr) return std::make_unique<ExprStmt>(std::move(expr));
    return nullptr;
}

std::unique_ptr<VarDeclStmt> Parser::parseVarDecl() {
    advance();
    std::string name = expect(TokenType::Ident).value;
    expect(TokenType::Eq);
    auto value = parseExpr();
    auto v = std::make_unique<VarDeclStmt>();
    v->name = name;
    v->value = std::move(value);
    return v;
}

std::unique_ptr<FuncDeclStmt> Parser::parseFuncDecl() {
    advance();
    std::string name = expect(TokenType::Ident).value;
    expect(TokenType::LParen);
    std::vector<std::string> params;
    while (!check(TokenType::RParen)) {
        if (check(TokenType::Ident) || check(TokenType::Self))
            params.push_back(advance().value);
        if (check(TokenType::Colon)) { advance(); expect(TokenType::Ident); }
        if (!check(TokenType::RParen)) expect(TokenType::Comma);
    }
    expect(TokenType::RParen);
    if (check(TokenType::Colon)) { advance(); expect(TokenType::Ident); }
    std::unique_ptr<ASTNode> body;
    if (check(TokenType::Arrow)) {
        advance();
        body = std::make_unique<ExprStmt>(parseExpr());
    } else {
        body = parseBlockRaw();
    }
    auto f = std::make_unique<FuncDeclStmt>();
    f->name = name;
    f->params = std::move(params);
    f->body = std::move(body);
    return f;
}

std::unique_ptr<ClassDeclStmt> Parser::parseClassDecl() {
    advance();
    std::string name = expect(TokenType::Ident).value;
    expect(TokenType::LBrace);
    auto c = std::make_unique<ClassDeclStmt>();
    c->name = name;
    while (!check(TokenType::RBrace) && !check(TokenType::Eof)) {
        if (check(TokenType::Pub) || check(TokenType::Priv)) {
            std::string vis = advance().value;
            std::string mname = expect(TokenType::Ident).value;
            std::unique_ptr<Expr> def;
            if (check(TokenType::Eq)) { advance(); def = parseExpr(); }
            c->members.emplace_back(vis, mname, std::move(def));
        } else if (check(TokenType::Fn)) {
            auto stmt = parseFuncDecl();
            c->methods.push_back(std::unique_ptr<FuncDeclStmt>(static_cast<FuncDeclStmt*>(stmt.release())));
        } else advance();
    }
    expect(TokenType::RBrace);
    return c;
}

std::unique_ptr<IfStmt> Parser::parseIf() {
    advance();
    auto cond = parseExpr();
    if (check(TokenType::LParen)) { advance(); expect(TokenType::RParen); }
    auto thenB = parseBlock();
    auto i = std::make_unique<IfStmt>();
    i->condition = std::move(cond);
    i->thenBlock = std::move(thenB);
    while (check(TokenType::Elif)) {
        advance();
        auto ec = parseExpr();
        auto eb = parseBlock();
        i->elifBlocks.emplace_back(std::move(ec), std::move(eb));
    }
    if (check(TokenType::Else)) {
        advance();
        i->elseBlock = parseBlock();
    }
    return i;
}

std::unique_ptr<ForStmt> Parser::parseFor() {
    advance();
    expect(TokenType::LParen);
    auto f = std::make_unique<ForStmt>();
    if (check(TokenType::Let)) {
        advance();
        std::string iterVar = expect(TokenType::Ident).value;
        if (check(TokenType::In)) {
            advance();
            f->iterVar = iterVar;
            f->iterExpr = parseExpr();
            expect(TokenType::RParen);
            f->body = parseBlock();
            return f;
        }
        expect(TokenType::Eq);
        auto v = std::make_unique<VarDeclStmt>();
        v->name = iterVar;
        v->value = parseExpr();
        f->init = std::move(v);
    } else if (check(TokenType::Ident)) {
        std::string n = advance().value;
        if (check(TokenType::In)) {
            advance();
            f->iterVar = n;
            f->iterExpr = parseExpr();
            expect(TokenType::RParen);
            f->body = parseBlock();
            return f;
        }
        if (check(TokenType::Eq)) {
            advance();
            auto a = std::make_unique<AssignStmt>();
            a->target = std::make_unique<IdentExpr>();
            static_cast<IdentExpr*>(a->target.get())->name = n;
            a->value = parseExpr();
            f->init = std::move(a);
        } else {
            auto e = std::make_unique<ExprStmt>();
            e->expr = std::make_unique<IdentExpr>();
            static_cast<IdentExpr*>(e->expr.get())->name = n;
            f->init = std::move(e);
        }
    } else if (!check(TokenType::Semicolon)) {
        f->init = parseStatement();
    }
    if (check(TokenType::Semicolon)) advance();
    if (!check(TokenType::Semicolon)) f->condition = parseExpr();
    expect(TokenType::Semicolon);
    if (!check(TokenType::RParen)) f->update = parseStatement();
    expect(TokenType::RParen);
    f->body = parseBlock();
    return f;
}

std::unique_ptr<WhileStmt> Parser::parseWhile() {
    advance();
    auto cond = parseExpr();
    if (check(TokenType::LParen)) { advance(); expect(TokenType::RParen); }
    auto w = std::make_unique<WhileStmt>();
    w->condition = std::move(cond);
    w->body = parseBlock();
    return w;
}

std::unique_ptr<BlockStmt> Parser::parseBlock() {
    expect(TokenType::LBrace);
    return parseBlockRaw();
}

std::unique_ptr<BlockStmt> Parser::parseBlockRaw() {
    auto b = std::make_unique<BlockStmt>();
    while (!check(TokenType::RBrace) && !check(TokenType::Eof)) {
        auto s = parseStatement();
        if (s) b->statements.push_back(std::move(s));
        if (check(TokenType::Semicolon)) advance();
    }
    expect(TokenType::RBrace);
    return b;
}

std::unique_ptr<Expr> Parser::parseExpr() { return parseOr(); }

std::unique_ptr<Expr> Parser::parseOr() {
    auto left = parseAnd();
    while (check(TokenType::Or)) {
        advance();
        auto b = std::make_unique<BinaryExpr>();
        b->left = std::move(left);
        b->op = "||";
        b->right = parseAnd();
        left = std::move(b);
    }
    return left;
}

std::unique_ptr<Expr> Parser::parseAnd() {
    auto left = parseEquality();
    while (check(TokenType::And)) {
        advance();
        auto b = std::make_unique<BinaryExpr>();
        b->left = std::move(left);
        b->op = "&&";
        b->right = parseEquality();
        left = std::move(b);
    }
    return left;
}

std::unique_ptr<Expr> Parser::parseEquality() {
    auto left = parseComparison();
    while (check(TokenType::EqEq) || check(TokenType::Neq)) {
        std::string op = advance().value;
        auto b = std::make_unique<BinaryExpr>();
        b->left = std::move(left);
        b->op = op;
        b->right = parseComparison();
        left = std::move(b);
    }
    return left;
}

std::unique_ptr<Expr> Parser::parseComparison() {
    auto left = parseTerm();
    while (check(TokenType::Lt) || check(TokenType::Le) || check(TokenType::Gt) || check(TokenType::Ge)) {
        std::string op = advance().value;
        auto b = std::make_unique<BinaryExpr>();
        b->left = std::move(left);
        b->op = op;
        b->right = parseTerm();
        left = std::move(b);
    }
    return left;
}

std::unique_ptr<Expr> Parser::parseTerm() {
    auto left = parseFactor();
    while (check(TokenType::Plus) || check(TokenType::Minus)) {
        std::string op = advance().value;
        auto b = std::make_unique<BinaryExpr>();
        b->left = std::move(left);
        b->op = op;
        b->right = parseFactor();
        left = std::move(b);
    }
    return left;
}

std::unique_ptr<Expr> Parser::parseFactor() {
    auto left = parseUnary();
    while (check(TokenType::Star) || check(TokenType::Slash) || check(TokenType::Percent)) {
        std::string op = advance().value;
        auto b = std::make_unique<BinaryExpr>();
        b->left = std::move(left);
        b->op = op;
        b->right = parseUnary();
        left = std::move(b);
    }
    return left;
}

std::unique_ptr<Expr> Parser::parseUnary() {
    if (check(TokenType::Minus) || check(TokenType::Not)) {
        std::string op = advance().value;
        auto u = std::make_unique<UnaryExpr>();
        u->op = op;
        u->right = parseUnary();
        return u;
    }
    return parseCall();
}

std::unique_ptr<Expr> Parser::parseCall() {
    auto base = parsePrimary();
    while (true) {
        if (check(TokenType::LParen)) {
            advance();
            auto c = std::make_unique<CallExpr>();
            c->callee = std::move(base);
            while (!check(TokenType::RParen)) {
                c->args.push_back(parseExpr());
                if (!check(TokenType::RParen)) expect(TokenType::Comma);
            }
            expect(TokenType::RParen);
            base = std::move(c);
        } else if (check(TokenType::LBracket)) {
            advance();
            auto idx = std::make_unique<IndexExpr>();
            idx->obj = std::move(base);
            idx->index = parseExpr();
            expect(TokenType::RBracket);
            base = std::move(idx);
        } else if (check(TokenType::Dot)) {
            advance();
            auto m = std::make_unique<MemberExpr>();
            m->obj = std::move(base);
            m->member = expect(TokenType::Ident).value;
            base = std::move(m);
        } else break;
    }
    return base;
}

std::unique_ptr<Expr> Parser::parsePrimary() {
    if (check(TokenType::New)) {
        advance();
        std::string cls = expect(TokenType::Ident).value;
        expect(TokenType::LParen);
        auto c = std::make_unique<CallExpr>();
        c->callee = std::make_unique<IdentExpr>();
        static_cast<IdentExpr*>(c->callee.get())->name = cls;
        while (!check(TokenType::RParen)) {
            c->args.push_back(parseExpr());
            if (!check(TokenType::RParen)) expect(TokenType::Comma);
        }
        expect(TokenType::RParen);
        return c;
    }
    if (check(TokenType::Number)) {
        auto n = std::make_unique<NumberExpr>();
        n->value = std::stod(advance().value);
        return n;
    }
    if (check(TokenType::String)) {
        auto s = std::make_unique<StringExpr>();
        s->value = advance().value;
        return s;
    }
    if (check(TokenType::True)) { advance(); auto b = std::make_unique<BoolExpr>(); b->value = true; return b; }
    if (check(TokenType::False)) { advance(); auto b = std::make_unique<BoolExpr>(); b->value = false; return b; }
    if (check(TokenType::Ident) || check(TokenType::Self)) {
        auto i = std::make_unique<IdentExpr>();
        i->name = advance().value;
        return i;
    }
    if (check(TokenType::LParen)) {
        advance();
        auto e = parseExpr();
        expect(TokenType::RParen);
        return e;
    }
    if (check(TokenType::LBracket)) {
        advance();
        auto a = std::make_unique<ArrayExpr>();
        while (!check(TokenType::RBracket)) {
            a->elements.push_back(parseExpr());
            if (!check(TokenType::RBracket)) expect(TokenType::Comma);
        }
        expect(TokenType::RBracket);
        return a;
    }
    if (check(TokenType::LBrace)) {
        advance();
        auto o = std::make_unique<ObjectExpr>();
        while (!check(TokenType::RBrace)) {
            std::string k = expect(TokenType::Ident).value;
            expect(TokenType::Colon);
            o->pairs.emplace_back(k, parseExpr());
            if (!check(TokenType::RBrace)) expect(TokenType::Comma);
        }
        expect(TokenType::RBrace);
        return o;
    }
    if (check(TokenType::Fn)) {
        advance();
        expect(TokenType::LParen);
        std::vector<std::string> params;
        while (!check(TokenType::RParen)) {
            if (check(TokenType::Ident) || check(TokenType::Self)) params.push_back(advance().value);
            if (check(TokenType::Colon)) { advance(); expect(TokenType::Ident); }
            if (!check(TokenType::RParen)) expect(TokenType::Comma);
        }
        expect(TokenType::RParen);
        std::unique_ptr<ASTNode> body;
        if (check(TokenType::Arrow)) { advance(); body = std::make_unique<ExprStmt>(parseExpr()); }
        else body = parseBlockRaw();
        auto l = std::make_unique<LambdaExpr>();
        l->params = std::move(params);
        l->body = std::move(body);
        return l;
    }
    throw ParseError("Unexpected token", current());
}

} // namespace mpp

File: mpp/parser.py
"""
M++ Parser — Builds AST from tokens
"""
from typing import List, Optional

from .lexer import Token, TokenType, Lexer
from .ast import (
    ASTNode, Program, Number, String, Bool, Identifier,
    BinaryOp, UnaryOp, Call, Index, Member, Array, Object,
    Conditional, Lambda,
    VarDecl, Assignment, ExprStmt, Block, IfStmt, ForStmt,
    WhileStmt, ReturnStmt, FuncDecl, ClassDecl,
)


class ParseError(Exception):
    def __init__(self, msg: str, token: Token):
        super().__init__(f"Line {token.line}: {msg}")
        self.token = token


class Parser:
    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.pos = 0

    def current(self) -> Token:
        if self.pos >= len(self.tokens):
            return self.tokens[-1]
        return self.tokens[self.pos]

    def advance(self) -> Token:
        t = self.current()
        if self.pos < len(self.tokens):
            self.pos += 1
        return t

    def expect(self, *types: TokenType) -> Token:
        t = self.current()
        if t.type not in types:
            raise ParseError(f"Expected {types}, got {t.type}", t)
        return self.advance()

    def check(self, *types: TokenType) -> bool:
        return self.current().type in types

    def parse(self) -> Program:
        stmts = []
        while not self.check(TokenType.EOF):
            stmt = self.parse_statement()
            if stmt:
                stmts.append(stmt)
            if self.check(TokenType.SEMICOLON):
                self.advance()
        return Program(stmts)

    def parse_statement(self) -> Optional[ASTNode]:
        if self.check(TokenType.LET, TokenType.CONST):
            return self.parse_var_decl()
        if self.check(TokenType.FN):
            return self.parse_func_decl()
        if self.check(TokenType.CLASS):
            return self.parse_class_decl()
        if self.check(TokenType.IF):
            return self.parse_if()
        if self.check(TokenType.FOR):
            return self.parse_for()
        if self.check(TokenType.WHILE):
            return self.parse_while()
        if self.check(TokenType.RETURN):
            self.advance()
            val = None if self.check(TokenType.SEMICOLON, TokenType.RBRACE, TokenType.EOF) else self.parse_expr()
            return ReturnStmt(val)
        if self.check(TokenType.LBRACE):
            return self.parse_block()

        expr = self.parse_expr()
        if expr and self.check(TokenType.EQ) and self._is_assignable(expr):
            self.advance()
            return Assignment(expr, self.parse_expr())
        return ExprStmt(expr) if expr else None

    def _is_assignable(self, node: ASTNode) -> bool:
        return isinstance(node, (Identifier, Index, Member))

    def parse_var_decl(self) -> VarDecl:
        mutable = self.advance().type == TokenType.LET
        name = self.expect(TokenType.IDENT).value
        type_hint = None
        if self.check(TokenType.COLON):
            self.advance()
            type_hint = self.expect(TokenType.IDENT).value
        self.expect(TokenType.EQ)
        value = self.parse_expr()
        return VarDecl(name, value, mutable, type_hint)

    def parse_func_decl(self) -> FuncDecl:
        self.expect(TokenType.FN)
        name = self.expect(TokenType.IDENT).value
        self.expect(TokenType.LPAREN)
        params = []
        while not self.check(TokenType.RPAREN):
            pname = self.expect(TokenType.IDENT, TokenType.SELF).value
            ptype = None
            if self.check(TokenType.COLON):
                self.advance()
                ptype = self.expect(TokenType.IDENT).value
            params.append((pname, ptype))
            if not self.check(TokenType.RPAREN):
                self.expect(TokenType.COMMA)
        self.expect(TokenType.RPAREN)
        ret_type = None
        if self.check(TokenType.COLON):
            self.advance()
            ret_type = self.expect(TokenType.IDENT).value
        if self.check(TokenType.ARROW):
            self.advance()
            body = ExprStmt(self.parse_expr())
        else:
            self.expect(TokenType.LBRACE)
            body = self.parse_block_raw()
        return FuncDecl(name, params, body, ret_type)

    def parse_class_decl(self) -> ClassDecl:
        self.expect(TokenType.CLASS)
        name = self.expect(TokenType.IDENT).value
        self.expect(TokenType.LBRACE)
        members = []
        methods = []
        while not self.check(TokenType.RBRACE, TokenType.EOF):
            if self.check(TokenType.PUB, TokenType.PRIV):
                vis = self.advance().value
                mname = self.expect(TokenType.IDENT).value
                val = None
                if self.check(TokenType.EQ):
                    self.advance()
                    val = self.parse_expr()
                members.append((vis, mname, val))
            elif self.check(TokenType.FN):
                methods.append(self.parse_func_decl())
            else:
                self.advance()
        self.expect(TokenType.RBRACE)
        return ClassDecl(name, members, methods)

    def parse_if(self) -> IfStmt:
        self.expect(TokenType.IF)
        cond = self.parse_expr()
        if self.check(TokenType.LPAREN):
            self.advance()
            self.expect(TokenType.RPAREN)
        then_block = self.parse_block()
        elif_branches = []
        while self.check(TokenType.ELIF):
            self.advance()
            elif_cond = self.parse_expr()
            elif_branch = self.parse_block()
            elif_branches.append((elif_cond, elif_branch))
        else_block = None
        if self.check(TokenType.ELSE):
            self.advance()
            else_block = self.parse_block()
        return IfStmt(cond, then_block, elif_branches, else_block)

    def parse_for(self) -> ForStmt:
        self.expect(TokenType.FOR)
        self.expect(TokenType.LPAREN)
        if self.check(TokenType.LET):
            self.advance()
            iter_var = self.expect(TokenType.IDENT).value
            if self.check(TokenType.IN):
                self.advance()
                iter_expr = self.parse_expr()
                self.expect(TokenType.RPAREN)
                body = self.parse_block()
                return ForStmt(None, None, None, body, iter_var, iter_expr)
            else:
                self.expect(TokenType.EQ)
                init = VarDecl(iter_var, self.parse_expr(), True, None)
        elif self.check(TokenType.IDENT):
            name_tok = self.advance()
            if self.check(TokenType.IN):
                iter_var = name_tok.value
                self.advance()
                iter_expr = self.parse_expr()
                self.expect(TokenType.RPAREN)
                body = self.parse_block()
                return ForStmt(None, None, None, body, iter_var, iter_expr)
            elif self.check(TokenType.EQ):
                self.advance()
                init = Assignment(Identifier(name_tok.value), self.parse_expr())
            else:
                init = ExprStmt(Identifier(name_tok.value))
        else:
            init = self.parse_statement() if not self.check(TokenType.SEMICOLON) else None
        if self.check(TokenType.SEMICOLON):
            self.advance()
        cond = None if self.check(TokenType.SEMICOLON) else self.parse_expr()
        self.expect(TokenType.SEMICOLON)
        update = None if self.check(TokenType.RPAREN) else self.parse_statement()
        self.expect(TokenType.RPAREN)
        body = self.parse_block()
        return ForStmt(init, cond, update, body)

    def parse_while(self) -> WhileStmt:
        self.expect(TokenType.WHILE)
        cond = self.parse_expr()
        if self.check(TokenType.LPAREN):
            self.advance()
            self.expect(TokenType.RPAREN)
        body = self.parse_block()
        return WhileStmt(cond, body)

    def parse_block(self) -> Block:
        self.expect(TokenType.LBRACE)
        return self.parse_block_raw()

    def parse_block_raw(self) -> Block:
        stmts = []
        while not self.check(TokenType.RBRACE, TokenType.EOF):
            stmt = self.parse_statement()
            if stmt:
                stmts.append(stmt)
            if self.check(TokenType.SEMICOLON):
                self.advance()
        self.expect(TokenType.RBRACE)
        return Block(stmts)

    def parse_expr(self) -> ASTNode:
        return self.parse_or()

    def parse_or(self) -> ASTNode:
        left = self.parse_and()
        while self.check(TokenType.OR):
            op = self.advance().value
            left = BinaryOp(left, op, self.parse_and())
        return left

    def parse_and(self) -> ASTNode:
        left = self.parse_equality()
        while self.check(TokenType.AND):
            op = self.advance().value
            left = BinaryOp(left, op, self.parse_equality())
        return left

    def parse_equality(self) -> ASTNode:
        left = self.parse_comparison()
        while self.check(TokenType.EQEQ, TokenType.NEQ):
            op = self.advance().value
            left = BinaryOp(left, op, self.parse_comparison())
        return left

    def parse_comparison(self) -> ASTNode:
        left = self.parse_term()
        while self.check(TokenType.LT, TokenType.LE, TokenType.GT, TokenType.GE):
            op = self.advance().value
            left = BinaryOp(left, op, self.parse_term())
        return left

    def parse_term(self) -> ASTNode:
        left = self.parse_factor()
        while self.check(TokenType.PLUS, TokenType.MINUS):
            op = self.advance().value
            left = BinaryOp(left, op, self.parse_factor())
        return left

    def parse_factor(self) -> ASTNode:
        left = self.parse_unary()
        while self.check(TokenType.STAR, TokenType.SLASH, TokenType.PERCENT):
            op = self.advance().value
            left = BinaryOp(left, op, self.parse_unary())
        return left

    def parse_unary(self) -> ASTNode:
        if self.check(TokenType.MINUS, TokenType.NOT):
            op = self.advance().value
            return UnaryOp(op, self.parse_unary())
        return self.parse_call()

    def parse_call(self) -> ASTNode:
        base = self.parse_primary()
        while True:
            if self.check(TokenType.LPAREN):
                self.advance()
                args = []
                while not self.check(TokenType.RPAREN):
                    args.append(self.parse_expr())
                    if not self.check(TokenType.RPAREN):
                        self.expect(TokenType.COMMA)
                self.expect(TokenType.RPAREN)
                base = Call(base, args)
            elif self.check(TokenType.LBRACKET):
                self.advance()
                idx = self.parse_expr()
                self.expect(TokenType.RBRACKET)
                base = Index(base, idx)
            elif self.check(TokenType.DOT):
                self.advance()
                member = self.expect(TokenType.IDENT).value
                base = Member(base, member)
            else:
                break
        return base

    def parse_primary(self) -> ASTNode:
        if self.check(TokenType.NEW):
            self.advance()
            class_name = self.expect(TokenType.IDENT).value
            self.expect(TokenType.LPAREN)
            args = []
            while not self.check(TokenType.RPAREN):
                args.append(self.parse_expr())
                if not self.check(TokenType.RPAREN):
                    self.expect(TokenType.COMMA)
            self.expect(TokenType.RPAREN)
            return Call(Identifier(class_name), args)
        if self.check(TokenType.NUMBER):
            return Number(self.advance().value)
        if self.check(TokenType.STRING):
            return String(self.advance().value)
        if self.check(TokenType.TRUE):
            self.advance()
            return Bool(True)
        if self.check(TokenType.FALSE):
            self.advance()
            return Bool(False)
        if self.check(TokenType.IDENT, TokenType.SELF):
            return Identifier(self.advance().value)
        if self.check(TokenType.LPAREN):
            self.advance()
            expr = self.parse_expr()
            self.expect(TokenType.RPAREN)
            return expr
        if self.check(TokenType.LBRACKET):
            self.advance()
            elements = []
            while not self.check(TokenType.RBRACKET):
                elements.append(self.parse_expr())
                if not self.check(TokenType.RBRACKET):
                    self.expect(TokenType.COMMA)
            self.expect(TokenType.RBRACKET)
            return Array(elements)
        if self.check(TokenType.LBRACE):
            self.advance()
            pairs = []
            while not self.check(TokenType.RBRACE):
                key = self.expect(TokenType.IDENT).value
                self.expect(TokenType.COLON)
                val = self.parse_expr()
                pairs.append((key, val))
                if not self.check(TokenType.RBRACE):
                    self.expect(TokenType.COMMA)
            self.expect(TokenType.RBRACE)
            return Object(pairs)
        if self.check(TokenType.FN):
            self.advance()
            self.expect(TokenType.LPAREN)
            params = []
            while not self.check(TokenType.RPAREN):
                params.append(self.expect(TokenType.IDENT, TokenType.SELF).value)
                if not self.check(TokenType.RPAREN):
                    self.expect(TokenType.COMMA)
            self.expect(TokenType.RPAREN)
            if self.check(TokenType.ARROW):
                self.advance()
                body = self.parse_expr()
            else:
                body = self.parse_block()
            return Lambda(params, body)
        raise ParseError("Unexpected token", self.current())
